### PR TITLE
fix contractor currentPeriod tracking bug

### PR DESCRIPTION
**Pull request by avahowell**
This PR fixes a bug in the contractor, where the allowance period was not being accurately tracked.

### CODE REVIEW

**Comment by aiorla** [comment]
I've also noticed that something wasn't right there today (I was trying to tackle #2308) and I said it on `#core-dev`. I was waiting for an answer to code it because I'm not sure about the expected behaviour.
I thought it was more like:
```go
if c.currentPeriod == 0 {
    c.currentPeriod = c.blockHeight
}
```
plus a few more lines to stop c.ProcessConsensusChange(...) to move it from 0.

**Comment by lukechampine** [comment]
merge conflicts after I merged the other PR

**Comment by avahowell** [comment]
fixed up the merge conflicts

File: modules/renter/contractor/allowance.go [hunk 1]
```diff
@@ -73,6 +73,7 @@ func (c *Contractor) SetAllowance(a modules.Allowance) error {
 	c.log.Println("INFO: setting allowance to", a)
 	c.mu.Lock()
 	c.allowance = a
+	c.currentPeriod = c.blockHeight
```
Comment (DavidVorick) [inline modules/renter/contractor/allowance.go]: This is not quite correct behavior I think. You don't want to reset the current period every time that you change the allowance. For example, if all you are doing here is adjusting the spending amount, you would expect the current period to not change.

I'm not actually certain what you should do if you are adjusting the period. If you increase the period, should you change the currentPeriod? If you decrase it, should you change the currentPeriod? Not clear to me at the moment what the best choice is there.

Comment (DavidVorick) [reply modules/renter/contractor/allowance.go]: Generally I think we want to follow a pattern of sticking to a rhythmic billing cycle, for example one a month, always on the same day. But if you change from a month to 6 weeks... I'm not sure what you are supposed to do.

And, does this code properly clean up the contract lines? I think that if you yank the current period around like this you can end up with contract lines that actually span multiple periods.

Comment (DavidVorick) [reply modules/renter/contractor/allowance.go]: My gut right now says we should only change the current period if `c.currentPeriod+(a.Period-a.RenewWindow) < c.blockHeight`, meaning that we would want to renew instantly.

Comment (avahowell) [reply modules/renter/contractor/allowance.go]: Would we also want to start the `currentPeriod` at the current blockheight if an allowance is set and there are no previous allowances?

Comment (DavidVorick) [reply modules/renter/contractor/allowance.go]: In most cases, that would satisfy the above equation, but I guess there are some corner cases if your `a.Period-a.RenewWindow` is greater than the block height, you could have issues.

So yeah, seems fine to check if the allowance is blank, and if so then you set the currentPeriod to the block height.

**Comment by DavidVorick** [comment]
@avahowell can you update this? Would like to see this merged into RC2.

**Comment by avahowell** [comment]
@DavidVorick updated

**Pull request merged**